src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1070,46 +1162,131 @@ DatabaseShardImp::findShardIndexToAdd(
+            {
+                JLOG(j_.error()) <<
+                    "Unable to finalize shard " << shardIndex;
+                return;
+            }
         }
-        avgShardFileSz_ = fileSz_ / complete_.size();
-    }
-    else
-        avgShardFileSz_ = 0;
 
-    if (incomplete_)
+        if (!shard->finalize(writeSQLite))
+        {
+            // Finalize failed, remove shard
+            {
+                std::lock_guard lock(mutex_);
+                shards_.erase(shardIndex);
+
+                using namespace boost::filesystem;
+                path const dir {shard->getDir()};
+                shard.reset();
+                try
+                {
+                    remove_all(dir);
+                }
+                catch (std::exception const& e)
+                {
+                    JLOG(j_.error()) << "exception " << e.what();
+                }
+
+                updateStatus(lock);
+            }
+
+            setFileStats();
+            return;
+        }
+
+        {
+            std::lock_guard lock(mutex_);
+            auto const it {shards_.find(shardIndex)};
+            if (it == shards_.end())
+                return;
+            it->second.state = ShardInfo::State::final;
+            updateStatus(lock);
+        }
+
+        setFileStats();
+
+        // Update peers with new shard index
+        if (!app_.config().standalone() &&
+            app_.getOPs().getOperatingMode() != OperatingMode::DISCONNECTED)
+        {
+            protocol::TMPeerShardInfo message;
+            PublicKey const& publicKey {app_.nodeIdentity().first};
+            message.set_nodepubkey(publicKey.data(), publicKey.size());
+            message.set_shardindexes(std::to_string(shardIndex));
+            app_.overlay().foreach(send_always(
+                std::make_shared<Message>(
+                    message,
+                    protocol::mtPEER_SHARD_INFO)));
+        }
+    });
+}
+
+void
+DatabaseShardImp::setFileStats()
+{
+    std::vector<std::weak_ptr<Shard>> wptrShards;
     {
-        fileSz_ += incomplete_->fileSize();
-        fdRequired_ += incomplete_->fdRequired();
+        std::lock_guard lock(mutex_);
+        assert(init_);
+
+        fileSz_ = 0;
+        fdRequired_ = 0;
+        avgShardFileSz_ = 0;
+
+        if (shards_.empty())
+            return;
+
+        for (auto const& e : shards_)
+            wptrShards.push_back(e.second.shard);
     }
 
-    if (!backed_)
-        return;
+    std::uint64_t sumSz {0};
+    std::uint32_t sumFd {0};
+    for (auto const& wptr : wptrShards)
+    {
+        if (auto shard {wptr.lock()}; shard)
+        {
+            auto[sz, fd] = shard->fileInfo();
+            sumSz += sz;
+            sumFd += fd;
+        }
+    }
 
-    // Require at least 15 file descriptors
-    fdRequired_ = std::max(fdRequired_, 15);
+    std::lock_guard lock(mutex_);
+    fileSz_ = sumSz;
+    fdRequired_ = sumFd;
+    avgShardFileSz_ = fileSz_ / wptrShards.size();
```

Comment:
Dividing by the size of the vector may result in inaccurate averages if some of the weak pointers were empty while aggregating the stats.

Reply:
Good find, will fix.